drivers/thermal/thermal_dummy.c

```diff
@@ -0,0 +1,348 @@
+static const struct cpufreq_frequency_table g_dummy_cpufreq_table[] =
+{
+  {100},
+  {300},
+  {500},
+  {700},
+  {900},
+  {CPUFREQ_TABLE_END},
+};
+static struct dummy_cpufreq_driver_s g_dummy_cpufreq_driver =
+{
+  .driver =
+    {
+      dummy_cpufreq_get_table,
+      dummy_cpufreq_target_index,
+      dummy_cpufreq_get_frequency,
+      dummy_cpufreq_suspend,
+      dummy_cpufreq_resume,
+    },
+  .table = g_dummy_cpufreq_table,
+  .table_len = nitems(g_dummy_cpufreq_table),
+};
+#endif /* CONFIG_THERMAL_DUMMY_CPUFREQ */
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/* Dummy Cooling Device Operations */
+
+static int dummy_cdev_set_state(FAR struct thermal_cooling_device_s *cdev,
+                                unsigned int state)
+{
+  FAR struct dummy_cooling_device_s *c = cdev->devdata;
+
+  c->cur_state = state;
+  return OK;
+}
+
+static int dummy_cdev_get_state(FAR struct thermal_cooling_device_s *cdev,
+                                FAR unsigned int *state)
+{
+  FAR struct dummy_cooling_device_s *c = cdev->devdata;
+
+  *state = c->cur_state;
+  return OK;
+}
+
+static int
+dummy_cdev_get_max_state(FAR struct thermal_cooling_device_s *cdev,
+                         FAR unsigned int *state)
+{
+  FAR struct dummy_cooling_device_s *c = cdev->devdata;
+
+  *state = c->max_state;
+  return OK;
+}
+
+/* Sensor */
+
+static int dummy_zdev_get_temp(FAR struct thermal_zone_device_s *zdev,
+                               FAR int *temp)
+{
+  FAR struct dummy_zone_device_s *s = zdev->devdata;
+
+  if (s->temperature >= DUMMY_TEMP_RANGE_HIGH)
+    {
+      s->raising = false;
+    }
+  else if (s->temperature <= DUMMY_TEMP_RANGE_LOW)
+    {
+      s->raising = true;
+    }
+
+  if (s->raising)
+    {
+      s->temperature++;
+    }
+  else
+    {
+      s->temperature--;
+    }
+
+  *temp = s->temperature;
+  return OK;
+}
+
+static int dummy_zdev_set_trips(FAR struct thermal_zone_device_s *zdev,
+                                int low, int high)
+{
+  return OK;
+}
+
+#ifdef CONFIG_THERMAL_DUMMY_CPUFREQ
+static FAR const struct cpufreq_frequency_table *dummy_cpufreq_get_table(
+                                           FAR struct cpufreq_policy *policy)
+{
+  FAR struct dummy_cpufreq_driver_s *driver =
+                         (FAR struct dummy_cpufreq_driver_s *)policy->driver;
+
+  return driver->table;
+}
+
+static int dummy_cpufreq_target_index(FAR struct cpufreq_policy *policy,
+                                      unsigned int index)
+{
+  FAR struct dummy_cpufreq_driver_s *driver =
+                         (FAR struct dummy_cpufreq_driver_s *)policy->driver;
+
+  DEBUGASSERT(index < driver->table_len);
+
+  driver->current.frequency = driver->table[index].frequency;
+  return 0;
+}
+
+static int dummy_cpufreq_get_frequency(FAR struct cpufreq_policy *policy)
```

Comment:
```suggestion
static int dummy_cpufreq_get_frequency(FAR struct cpufreq_policy *policy)
```
where is definition of cpufreq_policy

Reply:
> where is definition of cpufreq_policy

Sorry, the depends "cpufreq" has not been upstreamed yet, will fix depends later.

Reply:
"cpufreq" has not been upstreamed yet, commit 85ca089 fix the depends.
Selftest result(build and runtime) added to "Testing".